### DISCUSSION TITLE
IBM Equal Access Accessibility Checker violations

### DISCUSSION THREAD

**Discussion by luisrudge** [adobe/react-spectrum#7477] [General]
When running a page with a table through this checked, it highlights the following issue:

> Violation
> An explicit ARIA 'role' is not valid for <tr> element within an ARIA role 'grid' per the ARIA in HTML specification
> Table structure elements cannot specify an explicit 'role' within table containers

<img width="546" alt="image" src="https://github.com/user-attachments/assets/9e20a348-2dda-4700-a4a1-26019c6b1849">

Specifically, this happens because all the elements inside a table have a row, but this tool seems to think that this is invalid based on aria rules. Links provided by the checker:

https://www.ibm.com/able/requirements/requirements/#4_1_2
https://www.w3.org/TR/wai-aria-1.2/

<img width="1638" alt="image" src="https://github.com/user-attachments/assets/6824c893-47ad-4486-b84b-ce309e481f46">

Any thoughts on this?

**Comment by snowystinger** [comment]
@majornista

**Comment by majornista** [comment]
It seems this tool is validating against the ARIA in HTML specification: https://www.w3.org/TR/html-aria/#avoid-specifying-redundant-roles

> 3.2 Avoid specifying redundant roles
> 
> The following example illustrates a [button](https://html.spec.whatwg.org/multipage/form-elements.html#the-button-element) element which has also been provided an explicit role=button. Specifying this role is unnecessary, as a "button" element is already exposed with an implicit button role. **In practice this particular instance of redundancy will likely not have unforeseen side effects, other than unnecessarily making the markup more verbose, and incorrectly signaling to other authors that this practice is useful.** Please review the section [3.3 Be cautious of side effects](https://www.w3.org/TR/html-aria/#side-effects) for an example of where specifying unnecessary roles can be problematic.
> 
> [Example 4](https://www.w3.org/TR/html-aria/#example-redundant-role-on-button): Redundant role on button
> ```html
> 
> <button role="button">...</button>
> ```
> Similarly, the following uses a role=group on a [fieldset](https://html.spec.whatwg.org/multipage/form-elements.html#the-fieldset-element) element, and a role=Main on a [main](https://html.spec.whatwg.org/multipage/grouping-content.html#the-main-element) element. This is unnecessary, because the fieldset element is implicitly exposed as a role=group, as is the main element implicitly exposed as a role=main. **Again, in practice this will likely not have any unforeseen side effects to users of assistive technology, as long as the declaration of the role value uses [ASCII lowercase](https://infra.spec.whatwg.org/#ascii-lowercase).** Please see [4.4 Case requirements for ARIA role, state and property attributes](https://www.w3.org/TR/html-aria/#case-sensitivity) for more information.
> 
> [Example 5](https://www.w3.org/TR/html-aria/#example-redundant-role-on-fieldset-and-main): Redundant role on fieldset and main
> ```html
> 
> <fieldset role="group">...</fieldset>
> 
> <main role="Main">...</main>
> ```
> The following uses a role=list on an [ul](https://html.spec.whatwg.org/multipage/grouping-content.html#the-ul-element) element. As the ul element has an implicit role of list, explicitly adding the role would generally be considered redundant. However, some user agents suppress a list's implicit ARIA semantics if the list markers are removed from the visual presentation of the list items. Generally the redundant declaration of an element's implicit role would not be recommended, but in specific situations such as this, and where the role is necessary to expose, authors can explicitly add the role.
> 
> [Example 6](https://www.w3.org/TR/html-aria/#example-redundant-role-on-list): Redundant role on list
> ```html
> 
> <ul role="list">...</ul>
> ```

I checked with @jnurthen and we are in agreement that the IBM checker may be overstepping here by making a best practice recommendation a requirement.

Redundant, yet correct, aria roles have no perceivable effect on the behavior of assistive technologies.

I’m also a bit worried about the behavior when the table DOM structure is altered with virtualization.

That being said, we may be able to determine the implicit role of the element that the `useGridRow` hook returns, because it accepts a `ref` for the element as a parameter: https://github.com/adobe/react-spectrum/blob/369e1e77daa13bbb31b1bd1d8db2fd4b3d715c30/packages/%40react-aria/grid/src/useGridRow.ts#L42-L84

So at [line 70](https://github.com/adobe/react-spectrum/blob/369e1e77daa13bbb31b1bd1d8db2fd4b3d715c30/packages/%40react-aria/grid/src/useGridRow.ts#L70), use something like this:

```js
      role: (ref?.current?.tagName === 'TR' ? undefined : 'row'),
```

**Reply by luisrudge** [reply]
That's helpful context, thank you. I'm not advocating for this to be treated as a bug or anything (so no code changes required), but I wanted to hear your thoughts on this considering I'm no a11y expert. If you need anything else from me, please let me know.

**Reply by majornista** [reply]
Also, it should be noted that the ARIA in HTML document has a PR that updates its wording to explicitly allow, but not recommend, using a redundant role on table elements: https://github.com/w3c/html-aria/pull/436/files?diff=split&w=0#diff-0eb547304658805aad788d320f10bf1f292797b5e6d745a3bf617584da017051R3098-R3101

**Reply by luisrudge** [reply]
Considering it's not recommended (even if allowed), what's your take on removing this from the table component?

**Reply by majornista** [reply]
I think it might be better to open an issue against the IBM tool, which is elevating the severity level unnecessarily.

**Reply by majornista** [reply]
Note that in React-Aria, we replace all `tr`, `td`, and `th` elements with `div` elements when the collection is virtualized, so we have to use aria-* for that use case.

I actually spent some time yesterday experimenting with adding these conditionals to remove redundant aria-* attributes depending on the element being rendered, but there were still cases where the IBM testing tool was being way too strict, IMO. For example, it calls the use of `role="rowheader"` on a `td` as serious failure. It also objects to the omission of a `scope` attribute on a `th` with `role="colheader"`.

**Reply by luisrudge** [reply]
Sounds good. Thanks for the follow up!
